packages/next/server/node-web-streams-helper.ts

```diff
@@ -149,19 +149,44 @@ export function renderToInitialStream({
   return ReactDOMServer.renderToReadableStream(element, streamOptions)
 }
 
-export function createHeadInjectionTransformStream(
-  inject: () => Promise<string>
+function createHeadInsertionTransformStream(
+  insert: () => Promise<string>
 ): TransformStream<Uint8Array, Uint8Array> {
-  let injected = false
+  let inserted = false
+  let freezing = false
+  const queueTask =
+    process.env.NEXT_RUNTIME === 'edge' ? globalThis.setTimeout : setImmediate
   return new TransformStream({
     async transform(chunk, controller) {
       const content = decodeText(chunk)
       let index
-      if (!injected && (index = content.indexOf('</head')) !== -1) {
-        injected = true
-        const injectedContent =
-          content.slice(0, index) + (await inject()) + content.slice(index)
-        controller.enqueue(encodeText(injectedContent))
+      // While react is flushing chunks, we don't apply insertions
+      if (freezing) {
+        controller.enqueue(chunk)
+        return
+      }
+      if (inserted) {
+        // Insert the subsequent insertion content to end of body.
+        freezing = true
+        const insertion = await insert()
+        controller.enqueue(encodeText(insertion))
+        controller.enqueue(chunk)
+        // Queue a task to wait for react flushing chunks in microtasks,
+        // to avoid content being messed up with style tags together.
+        queueTask(() => {
+          freezing = false
+        })
+      } else if ((index = content.indexOf('</head')) !== -1) {
+        // When there's head element detected, insert into head element
+        freezing = true
+        const insertion = await insert()
+        const insertedHeadContent =
+          content.slice(0, index) + insertion + content.slice(index)
+        controller.enqueue(encodeText(insertedHeadContent))
+        inserted = true
+        queueTask(() => {
+          freezing = false
+        })
```

Comment:
You could probably hoist these closures out of the transform so that they're reusable instead of allocating a new one each time.